### PR TITLE
Lock updates support deleting & replacing reqs.

### LINKED ISSUE #2332: pex lock update: mechanism to remove projects from lockfile requirements
With `--project=` one can add a new project without changing the version of other entries, but there is not currently a mechanism to remove them in the same manner.  The lockfile needs to be recreated.

(I'd expect this to remove the "requirement" but not try to "force uninstall" the project from the lockfile.  That is if I have a lockfile with the requirements (`A>1.0.0`, `B<1.0.0`) and `A` depends on `B`.  I'd expect removing `B` to still leave whatever version was a dependency of `A` in the lock.)

### LINKED ISSUE #2334: pex lock update: mechanism to change the version specifier if an existing requirement (or constraint?)
With `--project=` one can adjust the current version of a project anywhere within the existing version specifier, but not change the specifier without regenerating the entire lockfile.  This would be useful for the same reasons as adding a new project with the lock is.

I'm unsure if constraints need separate consideration for changing the version specifier.  That is pip let's you mange them separately:
```
$ cat c.txt 
ansicolors<1.1.7
$ pip install -c c.txt  ansicolors==1.0.2
```
But I'm not sure if that means they must map to separate cli flags in Pex.

### PR DESCRIPTION
**Pull request by jsirois**
Previously you could only add new top-level requirements to a lock or
refine constraints against an existing locked item. Now you can `-d` /             `--delete-project` a top-level requirement from the lock as well as
replace a top-level requirement with `-R` / `--replace-project.

Fixes #2332                                                                        Fixes #2334

### CODE REVIEW

**Comment by jsirois** [comment]
Ok, another pretty big one. Thanks again for your review efforts.

The real changes are in:
+ [pex/cli/commands/lock.py](https://github.com/pantsbuild/pex/pull/2335/files#diff-e8d5b02d5703b5ab96946abe35a9b20be92e4d3a96d1bd5142c6ae58c4d60ca0): This is the CLI for `pex3 lock (update) ...` and performs post-processing of the update to display what would-be / is updated.
+ [pex/resolve/lockfile/updater.py](https://github.com/pantsbuild/pex/pull/2335/files#diff-c1af3f897d10df94f9e995bc04e7ead217ba540e53f5b7df8e4a9bb57465f716): This is the core lock update logic.
+ [tests/integration/cli/commands/test_lock_update_issues_2332_2334.py](https://github.com/pantsbuild/pex/pull/2335/files#diff-39b421061db753696c08ffde34eef78830e667b7926a496847e9cdd41ec16a29): The test of the new features.

File: pex/cli/commands/lock.py [hunk 1]
```diff
@@ -291,9 +292,35 @@ def _add_update_arguments(cls, update_parser):
...
+                "and is specified by a top-level requirement, the allowable updates will be "
+                "constrained by the original top-level requirement. In other words, for an "
+                "original top-level requirement of `requests>=2,<4` you could specify "
+                "`-p requests` to pick up new releases in the `>=2,<4` range specified by the "
+                "original top-level requirement or you could specify an overlapping range like "
+                "`-p 'requests<3.6,!==3.8.2'` to, for example, downgrade away from a newly "
+                "identified security vulnerability. If you specify a disjoint requirement like "
+                "`-p 'requests>=4'`, the operation will fail. If you really want to replace the "
+                "original top-level requirement and have this operation succeed, you can precede "
+                "the requirement with an `=` like so: `-p '=requests>=4'`. This option is mutually "
```
Comment (jsirois) [inline pex/cli/commands/lock.py:307]: I never really found any contingencies to force the choice between a new option, say `-r` / `--replace-project` and this `=`-prefix syntax on the existing `-p` / `--project` option like I did with `-d` / `--delete`. I'm ambivalent; so if you have some good reason this new replace should be split into its own flag, please speak up.

**Comment by jsirois** [comment]
One note to draw attention to is here: https://github.com/pantsbuild/pex/issues/2334#issuecomment-1897604016 

Basically, this change introduces everything needed to implement minimal lock auto-update, even for Pants. IOW ~like cargo where if you change the input requirements, the lock just gets auto-minimally updated implicitly. That said, it would be clunky to poke around in the existing lock to come up with the full set of `-p =<new-req-1> ... -p =<new-req-N> -d =<removed-proj-1> ... -d =<removed-proj-M>` to affect that style of update. This change was large; so I'd like to work through this PR and likely release, then circle back and add the lock update with bare requirements / `-r requirements` feature that will enable doing this automatically.

**Review by kaos — COMMENTED** [review]
only done a cursory read through..

File: pex/asserts.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+            pex_info = PexInfo.from_pex(pex)
+            pex_info.update(PexInfo.from_env())
+            assert_advice = "\n".join(
+                (assert_advice, "PEX-INFO:", json.dumps(pex_info.as_json_dict(), indent=2))
+            )
+        except Exception:
+            pass
+
+    raise AssertionError("\n".join((message, "---", assert_advice)))
```
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/asserts.py:60]: Yeah - that makes sense.

File: pex/cli/commands/lock.py (hunk 1 above)
Comment (jsirois) [reply pex/cli/commands/lock.py:307]: @kaos helped answer this here: https://github.com/pantsbuild/pex/issues/2332#issuecomment-1900318716

I'll fix by splitting out `-R` / `--replace-project` (`-r` is / will be taken by adding support for bare and `-r` requirements in a forthcoming PR).

File: pex/asserts.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsirois) [reply pex/asserts.py]: Ok, now fixed.

### When running from the Pex CLI (not inside a PEX):
```
$ python -c 'from pex.asserts import production_assert; production_assert(False, "Whoops!")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/asserts.py", line 67, in production_assert
    raise AssertionError("\n".join(message))
AssertionError: Whoops!
---
Pex 2.1.159
platform: Linux-5.15.90.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
argv: ['-c']
---
The error reported above resulted from an unexpected programming error which
you should never encounter.

Firstly, please accept our apology!

If you could file an issue with the error and details above, we'd be
grateful. You can do that at https://github.com/pantsbuild/pex/issues/new and
redact or amend any details that expose sensitive information.
```

### When running from inside a PEX:
```diff
diff --git a/pex/pex_bootstrapper.py b/pex/pex_bootstrapper.py
index 5fb1ce5b..e9eefe3a 100644
--- a/pex/pex_bootstrapper.py
+++ b/pex/pex_bootstrapper.py
@@ -589,6 +589,10 @@ def bootstrap_pex(
 ):
     # type: (...) -> None

+    from pex.asserts import production_assert
+
+    production_assert(False, "Whoops!")
+
     pex_info = _bootstrap(entry_point)

     # ENV.PEX_ROOT is consulted by PythonInterpreter and Platform so set that up as early as
```

```
$ python -mpex -oempty.pex
jsirois@Gill-Windows:~/dev/pantsbuild/jsirois-pex (issues/2332 *) $ ./empty.pex
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/jsirois/.pex/unzipped_pexes/1d14ae7381aa17f6fbe3bb38e94143068a89091f/__main__.py", line 106, in <module>
    bootstrap_pex(__entry_point__, execute=__execute__, venv_dir=__venv_dir__)
  File "/home/jsirois/.pex/unzipped_pexes/1d14ae7381aa17f6fbe3bb38e94143068a89091f/.bootstrap/pex/pex_bootstrapper.py", line 594, in bootstrap_pex
    production_assert(False, "Whoops!")
  File "/home/jsirois/.pex/unzipped_pexes/1d14ae7381aa17f6fbe3bb38e94143068a89091f/.bootstrap/pex/asserts.py", line 67, in production_assert
    raise AssertionError("\n".join(message))
AssertionError: Whoops!
---
Pex 2.1.159
platform: Linux-5.15.90.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
argv: ['/home/jsirois/dev/pantsbuild/jsirois-pex/./empty.pex']
PEX-INFO:
{
  "bootstrap_hash": "8b3fa58512e1e46fbf845a9258d0fc83af041aff",
  "build_properties": {
    "pex_version": "2.1.159"
  },
  "code_hash": "c7af8ab820d0b72cc617d52b81c7ee6da79d1858",
  "deps_are_wheel_files": false,
  "distributions": {},
  "emit_warnings": true,
  "excluded": [],
  "ignore_errors": false,
  "includes_tools": false,
  "inherit_path": "false",
  "inject_args": [],
  "inject_env": {},
  "interpreter_constraints": [],
  "max_install_jobs": 1,
  "pex_hash": "1d14ae7381aa17f6fbe3bb38e94143068a89091f",
  "pex_path": "",
  "pex_paths": [],
  "requirements": [],
  "strip_pex_env": true,
  "venv": false,
  "venv_bin_path": "false",
  "venv_copies": false,
  "venv_hermetic_scripts": true,
  "venv_site_packages_copies": false
}
---
The error reported above resulted from an unexpected programming error which
you should never encounter.

Firstly, please accept our apology!

If you could file an issue with the error and details above, we'd be
grateful. You can do that at https://github.com/pantsbuild/pex/issues/new and
redact or amend any details that expose sensitive information.
```

**Review by benjyw — APPROVED** [review]
Did a non-cursory review, and it all looks good to me. I've verified via eyeballing that this should work in the presence of user-provided inconsistently non-canonical project names. 

Users will love this (post https://github.com/pantsbuild/pex/issues/2334#issuecomment-1897604016 anyway). And once we have that, Pants should be able to auto-update lockfiles when requirements change, a la Cargo.

**Comment by jsirois** [comment]
> Users will love this (post https://github.com/pantsbuild/pex/issues/2334#issuecomment-1897604016 anyway). And once we have that, Pants should be able to auto-update lockfiles when requirements change, a la Cargo.

As that comment alluded, nothing blocks that functionality as soon as this is merged. Its clunky for a human, but Pants is not that. IIUC it is easy for pants to create a lock if none present, or else update the present lock by gathering requirements, compare to prior requirements in lock, separate in 2 groups of changed + new and deleted, call pex3 lock update with the corresponding set of -R and -d. That said, I can see if you want to wait instead of writing a bunch of logic you then turn around and rip out shortly after.

**Pull request merged**